src/System.Text.Json/src/System/Text/Json/Serialization/MemberAccessor.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json.Serialization
+{
+    internal static class MemberAccessor
+    {
+        private delegate TProperty GetProperty<TClass, TProperty>(TClass obj);
+        private delegate TProperty GetPropertyByRef<TClass, TProperty>(ref TClass obj);
+
+        private delegate void SetProperty<TClass, TProperty>(TClass obj, TProperty value);
+        private delegate void SetPropertyByRef<TClass, TProperty>(ref TClass obj, TProperty value);
+
+        private delegate Func<object, TProperty> GetPropertyByRefFactory<TClass, TProperty>(GetPropertyByRef<TClass, TProperty> set);
+        private delegate Action<object, TProperty> SetPropertyByRefFactory<TClass, TProperty>(SetPropertyByRef<TClass, TProperty> set);
+
+        private static readonly MethodInfo s_createStructPropertyGetterMethod = new GetPropertyByRefFactory<int, int>(CreateStructPropertyGetter)
+            .Method.GetGenericMethodDefinition();
+
+        private static readonly MethodInfo s_createStructPropertySetterMethod = new SetPropertyByRefFactory<int, int>(CreateStructPropertySetter)
+            .Method.GetGenericMethodDefinition();
+
+        internal static Func<object, TProperty> CreatePropertyGetter<TClass, TProperty>(PropertyInfo propertyInfo)
+        {
+            MethodInfo getMethodInfo = propertyInfo.GetGetMethod();
+
+            if (typeof(TClass).IsValueType)
+            {
+                var factory = CreateDelegate<GetPropertyByRefFactory<TClass, TProperty>>(s_createStructPropertyGetterMethod.MakeGenericMethod(typeof(TClass), typeof(TProperty)));
+                var propertyGetter = CreateDelegate<GetPropertyByRef<TClass, TProperty>>(getMethodInfo);
+
+                return factory(propertyGetter);
+            }
+            else
+            {
+                var propertyGetter = CreateDelegate<GetProperty<TClass, TProperty>>(getMethodInfo);
```

Comment:
Concerned about performance here. Have you measured? 

Some comments here would be good as well - like the double hop.

I believe you need both a strongly-typed class (TClass) and property (TProperty) in order to have a direct reference to the getter\setter (instead of methodinfo.invoke) which we have above however then this new delegate below is returned which doesn't have a strongly-typed TClass.

Also concerned about the extra hop (assuming there is one).

Reply:
It's slower (not measured, but it is) because of the second delegate one line below. Unfortunately, it's a price of `Unsafe.Unbox<T>` which has a constraint on `T`. I will think about merging two generic `JsonPropertyInfo` into a single class and about adding specialized classes for struct and class handling.

> this new delegate below is returned which doesn't have a strongly-typed TClass.

Yep, because passing a value as `TClass` when it's a structure will cause passing it by value, not by reference. To avoid it I pass an object and get a reference to it in lines 80 and 89. As I said, there is a constraint on `T` which can be avoided by some magic with delegates. The JIT doesn't check constrains, but the compiler does,

Reply:
Can we can avoid the second delegate for the reference type (non-struct) case?

Reply:
Nope due to the signature check.

Reply:
Let me benchmark this and see what the hit is.